src/modules/core/filter_pillar_echo.c

```diff
@@ -0,0 +1,314 @@
+{
+	scale_sliced_desc desc;
+	desc.src = src;
+	desc.dst = dst;
+	desc.rect = rect;
+
+	// Crop out a section of the rect that has the same aspect ratio as the image
+	double destAr = (double)src->width / (double)src->height;
+	double sourceAr = rect.w / rect.h;
+	if ( sourceAr > destAr )
+	{
+		// Crop sides to fit height
+		desc.rect.w = rect.w * destAr / sourceAr;
+		desc.rect.x = rect.x + (rect.w - desc.rect.w) / 2.0;
+	}
+	else if ( destAr > sourceAr )
+	{
+		// Crop top and bottom to fit width.
+		desc.rect.h = rect.h * sourceAr / destAr;
+		desc.rect.y = rect.y + (rect.h - desc.rect.h) / 2.0;
+	}
+	mlt_slices_run_normal(0, scale_sliced_proc, &desc);
+}
+
+
+/** Copy pixels from source to destination
+  *
+  * \param src a pointer to the source image
+  * \param dst a pointer to the destination image
+  * \param width the number of values in each row
+  * \param rect the area of interest in the src to be copied to the dst
+  */
+
+static void blit_rect( mlt_image src, mlt_image dst, mlt_rect rect )
+{
+	int blitHeight = rect.h;
+	int blitWidth = rect.w * 4;
+	int linesize = src->width * 4;
+	uint8_t* s = src->data + (int)rect.y * linesize + (int)rect.x * 4;
+	uint8_t* d = dst->data + (int)rect.y * linesize + (int)rect.x * 4;
+	while ( blitHeight-- )
+	{
+		memcpy( d, s, blitWidth );
+		s += linesize;
+		d += linesize;
+	}
+}
+
+static int filter_get_image( mlt_frame frame, uint8_t **image, mlt_image_format *format, int *width, int *height, int writable )
+{
+	int error = 0;
+	mlt_filter filter = (mlt_filter)mlt_frame_pop_service( frame );
+	mlt_properties filter_properties = MLT_FILTER_PROPERTIES( filter );
+	char* rect_str = mlt_properties_get( filter_properties, "rect" );
+	if ( !rect_str )
+	{
+		mlt_log_warning( MLT_FILTER_SERVICE(filter), "rect property not set\n" );
+		return mlt_frame_get_image( frame, image, format, width, height, writable );
+	}
+	mlt_profile profile = mlt_service_profile( MLT_FILTER_SERVICE( filter ) );
+	mlt_position position = mlt_filter_get_position( filter, frame );
+	mlt_position length = mlt_filter_get_length2( filter, frame );
+	mlt_rect rect = mlt_properties_anim_get_rect( filter_properties, "rect", position, length );
+	if ( strchr( rect_str, '%' ) )
+	{
+		rect.x *= profile->width;
+		rect.w *= profile->width;
+		rect.y *= profile->height;
+		rect.h *= profile->height;
+	}
+	double scale = mlt_profile_scale_width( profile, *width );
+	rect.x *= scale;
+	rect.w *= scale;
+	scale = mlt_profile_scale_height( profile, *height );
+	rect.y *= scale;
+	rect.h *= scale;
+	rect = constrain_rect( rect, profile->width * scale, profile->height * scale );
+
+	if ( rect.w < 1 || rect.h < 1 )
+	{
+		mlt_log_info( MLT_FILTER_SERVICE(filter), "rect invalid\n" );
+		return mlt_frame_get_image( frame, image, format, width, height, writable );
+	}
+
+	*format = mlt_image_rgba;
+	error = mlt_frame_get_image( frame, image, format, width, height, 1 );
```

Comment:
Here you are requesting a writable image from upstream, but that is not required. That can cause some producers to be slightly less efficient.

Reply:
Good to know. Fixed!